ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -535,40 +567,54 @@ private PlanNode buildSourceNode(final boolean isWindowed) {
    * @return the PlanNode representing this Join Tree
    */
   private JoinNode buildJoin(final Join root, final String prefix, final boolean isWindowed) {
-    final PlanNode left;
+    final PlanNode preRepartitionLeft;
     if (root.getLeft() instanceof JoinTree.Join) {
-      left = buildSourceForJoin(
-          (JoinTree.Join) root.getLeft(),
-          buildJoin((Join) root.getLeft(), prefix + "L_", isWindowed),
-          prefix + "Left",
-          root.getInfo().getLeftJoinExpression()
-      );
+      preRepartitionLeft = buildJoin((Join) root.getLeft(), prefix + "L_", isWindowed);
     } else {
       final JoinTree.Leaf leaf = (Leaf) root.getLeft();
-      left = buildSourceForJoin(
-          leaf.getSource(), prefix + "Left", root.getInfo().getLeftJoinExpression(), isWindowed);
+      preRepartitionLeft = new DataSourceNode(
+          new PlanNodeId("KafkaTopic_" + prefix + "Left"),
+          leaf.getSource().getDataSource(),
+          leaf.getSource().getAlias(),
+          isWindowed
+      );
     }
 
-    final PlanNode right;
+    final PlanNode preRepartitionRight;
     if (root.getRight() instanceof JoinTree.Join) {
-      right = buildSourceForJoin(
-          (JoinTree.Join) root.getRight(),
-          buildJoin((Join) root.getRight(), prefix + "R_", isWindowed),
-          prefix + "Right",
-          root.getInfo().getRightJoinExpression()
-      );
+      preRepartitionRight = buildJoin((Join) root.getRight(), prefix + "R_", isWindowed);
     } else {
       final JoinTree.Leaf leaf = (Leaf) root.getRight();
-      right = buildSourceForJoin(
-          leaf.getSource(), prefix + "Right", root.getInfo().getRightJoinExpression(), isWindowed);
+      preRepartitionRight = new DataSourceNode(
+          new PlanNodeId("KafkaTopic_" + prefix + "Right"),
+          leaf.getSource().getDataSource(),
+          leaf.getSource().getAlias(),
+          isWindowed
+      );
     }
 
-    verifyJoin(root.getInfo(), left, right);
+    final boolean isForeignKeyJoin =
+        verifyJoin(root.getInfo(), preRepartitionLeft, preRepartitionRight);
 
     final boolean finalJoin = prefix.isEmpty();
 
     final JoinKey joinKey = buildJoinKey(root);
 
+    final PlanNode left = prepareSourceForJoin(
```

Comment:
Wondering if we should pass the `isForeignKeyJoin` into this method, or do the check right here instead?
```
final PlanNode left;
final PlanNode right;
if (isForeignKeyJoin) {
  left = preRepartitionLeft;
  right = preRepartitionRight;
} else {
  left = prepareSourceForJoin(...);
  right = left = prepareSourceForJoin(...);
}
```

The idea is to keep the logic "contained". I seems `prepareSourceForJoin` does not really need to know anything about FK-joins? Atm, it feels like as if we have a "leaky" abstraction.

Reply:
If we have a foreign key join, then we don't want PreJoinRepartition nodes but we do still want PreJoinProject nodes on top of DataSourceNodes (the non-nested-join case). Therefore the logic in the foreign key join case is not as simple as in your sample code above, and it makes sense to continue wrapping it in a helper method. (Also helps de-dup code from the nested-join case since the logic for building PreJoinProject nodes is the same in both cases.)

Reply:
I see. I guess, it's good enough for now.

But it seems that `PreJoinProject` is actually an optimization, and thus should (mid term) not be part of the logical planner. The logical planner should generate a straight-forward plan and just verify if the query is valid and can be planned. This logical plan would be the input to an optimizer that modify the plan into an optimized (logical) plan, including projection push-down. More follow up refactoring for the future :)